File: frontend/src/types/messages.ts
export interface AttachmentRead {
  attachment_id: string
  file_name: string
  extension: string | null
  content_type: string | null
  size_bytes: number | null
  file_date: string | null
  matches_naming_convention: boolean
  matches_search_pattern: boolean | null
  content_sha256: string | null
}

export interface AttachmentListItem {
  attachment_id: string
  message_id: string
  file_name: string
  extension: string | null
  content_type: string | null
  size_bytes: number | null
  file_date: string | null
  matches_naming_convention: boolean
  matches_search_pattern: boolean | null
  content_sha256: string | null
  content_sha256_computed_at: string | null
  message_subject: string | null
  message_from_address: string | null
  message_sent_datetime: string | null
  folder_path: string | null
  mailbox_account_id: number | null
  mailbox_label: string | null
  linked_to_case: boolean
}

export interface MessageListItem {
  message_id: string
  conversation_id: string | null
  subject: string | null
  from_address: string | null
  from_name: string | null
  sent_datetime: string | null
  has_attachments: boolean
  is_sent: boolean
  folder_id: string | null
  folder_path: string | null
  mailbox_account_id: number | null
  mailbox_label: string | null
  attachments: AttachmentRead[]
}

export interface MessageDetail {
  message_id: string
  conversation_id: string | null
  internet_message_id: string | null
  subject: string | null
  from_address: string | null
  from_name: string | null
  to_addresses: string[]
  cc_addresses: string[]
  sent_datetime: string | null
  received_datetime: string | null
  has_attachments: boolean
  importance: string | null
  is_sent: boolean
  categories: string[]
  body_preview: string | null
  body_content: string | null
  body_content_type: string
  web_link: string | null
  folder_id: string | null
  folder_path: string | null
  mailbox_account_id: number | null
  mailbox_label: string | null
  attachments: AttachmentRead[]
}

export interface ConversationRead {
  conversation_id: string
  message_count: number
  first_message_at: string | null
  last_message_at: string | null
  participants: string[]
  messages: MessageListItem[]
}

export interface MailFolderNode {
  folder_id: string
  parent_folder_id: string | null
  display_name: string
  folder_path: string | null
  child_folder_count: number
  total_item_count: number
  last_sync_at: string | null
  children: MailFolderNode[]
}

export interface MessageFilters {
  folder_id?: string
  date_from?: string
  date_to?: string
  from_address?: string
  subject_contains?: string
  text_search?: string
  text_contains?: string
  has_attachments?: boolean
  attachment_pattern?: string
  mailbox_account_id?: number
  limit?: number
  offset?: number
}

File: backend/app/schemas/messages.py
from datetime import date, datetime

from pydantic import BaseModel


class AttachmentRead(BaseModel):
    attachment_id: str
    file_name: str
    extension: str | None
    content_type: str | None
    size_bytes: int | None
    file_date: date | None
    matches_naming_convention: bool
    matches_search_pattern: bool | None
    content_sha256: str | None = None


class AttachmentListItem(BaseModel):
    attachment_id: str
    message_id: str
    file_name: str
    extension: str | None
    content_type: str | None
    size_bytes: int | None
    file_date: date | None
    matches_naming_convention: bool
    matches_search_pattern: bool | None
    content_sha256: str | None
    content_sha256_computed_at: datetime | None
    message_subject: str | None
    message_from_address: str | None
    message_sent_datetime: datetime | None
    folder_path: str | None
    mailbox_account_id: int | None
    mailbox_label: str | None
    linked_to_case: bool


class MessageListItem(BaseModel):
    message_id: str
    conversation_id: str | None
    subject: str | None
    from_address: str | None
    from_name: str | None
    sent_datetime: datetime | None
    has_attachments: bool
    is_sent: bool
    folder_id: str | None
    folder_path: str | None
    mailbox_account_id: int | None
    mailbox_label: str | None
    attachments: list[AttachmentRead] = []


class MessageDetail(BaseModel):
    message_id: str
    conversation_id: str | None
    internet_message_id: str | None
    subject: str | None
    from_address: str | None
    from_name: str | None
    to_addresses: list[str]
    cc_addresses: list[str]
    sent_datetime: datetime | None
    received_datetime: datetime | None
    has_attachments: bool
    importance: str | None
    is_sent: bool
    categories: list[str]
    body_preview: str | None
    body_content: str | None
    body_content_type: str
    web_link: str | None
    folder_id: str | None
    folder_path: str | None
    mailbox_account_id: int | None
    mailbox_label: str | None
    attachments: list[AttachmentRead]


class ConversationRead(BaseModel):
    conversation_id: str
    message_count: int
    first_message_at: datetime | None
    last_message_at: datetime | None
    participants: list[str]
    messages: list[MessageListItem]


class RetraceAttachmentsResponse(BaseModel):
    traced_count: int


class MailFolderRead(BaseModel):
    folder_id: str
    parent_folder_id: str | None
    display_name: str
    folder_path: str | None
    child_folder_count: int
    total_item_count: int
    last_sync_at: datetime | None


class MailFolderNode(MailFolderRead):
    children: list["MailFolderNode"] = []


MailFolderNode.model_rebuild()
